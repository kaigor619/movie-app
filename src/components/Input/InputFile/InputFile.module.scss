.inputFileLabel {
  padding: 0.6rem 0rem 0.6rem 0rem;
  color: rgba(var(--text-primary-rgb), 0.6);
  border-radius: 0.4rem;
  font-size: 0.85rem;
  width: 100%;
  display: block;
  box-sizing: border-box;
  text-align: center;
  cursor: pointer;
  display: flex;
  align-items: center;
  justify-content: center;
  line-height: normal;
  transition: 0.2s color;

  & > svg {
    // color: rgb(255 255 255 / 70%);
    width: 1.1rem;
    height: auto;
  }

  & > span {
    margin-left: 0.4rem;
    margin-top: 0.1rem;
  }

  &:hover {
    color: rgba(var(--text-primary-rgb), 0.9);
  }

  &.active {
    color: rgba(var(--text-primary-rgb), 0.9);
  }
}

.inputFile {
  display: none;
}
