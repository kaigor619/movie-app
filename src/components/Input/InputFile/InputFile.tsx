import { IconDownload } from "@tabler/icons";
import clsx from "clsx";
import styles from "./InputFile.module.scss";

interface Props {
  name: string;
  value: File | null;
  onChange: (e: File | null) => void;
}

export const InputFile: React.FC<Props> = ({ value, name, onChange }) => {
  return (
    <>
      <label
        className={clsx(styles.inputFileLabel, {
          [styles.active]: !!value,
        })}
        htmlFor={name}
      >
        <IconDownload />
        <span>{value ? value.name : "Choose image"}</span>
      </label>

      <input
        id={name}
        type="file"
        className={styles.inputFile}
        onChange={(e) => onChange(e.target.files ? e.target.files[0] : null)}
      />
    </>
  );
};
