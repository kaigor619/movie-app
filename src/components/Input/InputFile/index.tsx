export * from "./InputFile";
export * from "./InputFileController";
