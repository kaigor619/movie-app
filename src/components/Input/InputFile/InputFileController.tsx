import { Controller, UseControllerProps } from "react-hook-form";

import { InputFile } from "./InputFile";

// type InputProps = React.ComponentProps<typeof InputFile>;

type Props<T extends Record<string, any>> = RequiredBy<
  UseControllerProps<T>,
  "control"
>;

export function InputFileController<T extends Record<string, any>>(
  props: Props<T>
): React.ReactElement<Props<T>> {
  return (
    <Controller
      name={props.name}
      control={props.control}
      rules={props.rules}
      render={({ field }) => (
        <InputFile
          name={props.name}
          value={field.value}
          onChange={field.onChange}
        />
      )}
    />
  );
}
