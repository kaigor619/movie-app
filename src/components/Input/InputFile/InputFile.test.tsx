import { screen, render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { InputFile } from "./InputFile";

describe("InputFile component", () => {
  const onChange = jest.fn();
  const file = new File([""], "filename.txt", {
    type: "text/plain",
    lastModified: Date.now(),
  });

  afterEach(() => {
    onChange.mockClear();
  });

  test("shows label without value", () => {
    render(<InputFile name="file" value={null} onChange={onChange} />);
    expect(screen.getByText(/Choose image/i)).toBeInTheDocument();
  });
  test("shows file name", () => {
    render(<InputFile name="file" value={file} onChange={onChange} />);

    expect(screen.getByText(file.name)).toBeInTheDocument();
  });
  test("checks snapshot", () => {
    const { asFragment } = render(
      <InputFile name="file" value={file} onChange={onChange} />
    );

    expect(asFragment()).toMatchSnapshot();
  });
  test("onChange", async () => {
    render(<InputFile name="file" value={null} onChange={onChange} />);

    await userEvent.upload(screen.getByTestId("input-file"), file);

    expect(onChange).toHaveBeenCalledWith(file);
    expect(onChange).toHaveBeenCalledTimes(1);
  });
});
