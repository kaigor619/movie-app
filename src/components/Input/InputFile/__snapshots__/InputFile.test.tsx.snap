// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputFile component checks snapshot 1`] = `
<DocumentFragment>
  <label
    class="inputFileLabel active"
    for="file"
  >
    <svg
      class="icon icon-tabler icon-tabler-download"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M0 0h24v24H0z"
        fill="none"
        stroke="none"
      />
      <path
        d="M4 17v2a2 2 0 0 0 2 2h12a2 2 0 0 0 2 -2v-2"
      />
      <polyline
        points="7 11 12 16 17 11"
      />
      <line
        x1="12"
        x2="12"
        y1="4"
        y2="16"
      />
    </svg>
    <span>
      filename.txt
    </span>
  </label>
  <input
    class="inputFile"
    data-testid="input-file"
    id="file"
    name="file"
    type="file"
  />
</DocumentFragment>
`;
